app/packages/core/src/components/Grid/useSpotlightPager.ts

```diff
@@ -0,0 +1,129 @@
+  schema: Schema,
+  zoom: boolean,
+  records: Map<string, number>
+) => {
+  return data.samples.edges.map((edge, i) => {
+    if (edge.node.__typename === "%other") {
+      throw new Error("unexpected sample type");
+    }
+
+    const id = { description: edge.node.id };
+    store.set(id, edge.node);
+    records.set(edge.node.id, page * PAGE_SIZE + i);
+
+    return {
+      key: page,
+      aspectRatio: zoom
+        ? zoomAspectRatio(edge.node.sample, schema, edge.node.aspectRatio)
+        : edge.node.aspectRatio,
+      id,
+      data: edge.node as fos.Sample,
+    };
+  });
+};
+
+const useSpotlightPager = ({
+  pageSelector,
+  zoomSelector,
+}: {
+  pageSelector: RecoilValueReadOnly<
+    (page: number, pageSize: number) => VariablesOf<foq.paginateSamplesQuery>
+  >;
+  zoomSelector: RecoilValueReadOnly<boolean>;
+}) => {
+  const environment = useRelayEnvironment();
+  const pager = useRecoilValue(pageSelector);
+  const zoom = useRecoilValue(zoomSelector);
+  const handleError = useErrorHandler();
+  const store = useMemo(
+    () => new WeakMap<ID, { sample: fos.Sample; index: number }>(),
+    []
+  );
+  const records = useRef(new Map<string, number>());
+
+  const page = useRecoilCallback(
+    ({ snapshot }) => {
+      return async (pageNumber: number) => {
+        const variables = pager(pageNumber, PAGE_SIZE);
+        let subscription: Subscription;
+        const schema = await snapshot.getPromise(
+          fos.fieldSchema({ space: fos.State.SPACE.SAMPLE })
+        );
+        return new Promise<Response<number, fos.Sample>>((resolve) => {
+          subscription = fetchQuery<foq.paginateSamplesQuery>(
+            environment,
+            foq.paginateSamples,
+            variables,
+            {
+              networkCacheConfig: { metadata: {} },
+              fetchPolicy: "store-or-network",
+            }
+          ).subscribe({
+            next: (data) => {
+              const items = processSamplePageData(
+                pageNumber,
+                store,
+                data,
+                schema,
+                zoom,
+                records.current
+              );
+
+              resolve({
+                items,
+                next: data.samples.pageInfo.hasNextPage ? pageNumber + 1 : null,
+                previous: pageNumber > 0 ? pageNumber - 1 : null,
+              });
+            },
+            complete: () => {
+              subscription?.unsubscribe();
+            },
+            error: handleError,
+          });
+        });
+      };
+    },
+    [environment, handleError, pager, store, zoom]
+  );
+
+  const refresher = useRecoilValue(fos.refresher);
+  useEffect(() => {
+    refresher;
+    const current = records.current;
+    return () => {
+      commitLocalUpdate(fos.getCurrentEnvironment(), (store) => {
+        for (const id of Array.from(current.keys())) {
+          store.get(id)?.invalidateRecord();
+        }
+        current.clear();
```

Comment:
On refresher changes, e.g. from tagging, invalidate relay records so new network requests for samples are issued

Reply:
same comment as above, would love to see these comments in the code